### ISSUE TITLE
Top-level if statement unexpectedly removed.

### ISSUE THREAD

**Issue by mbostock** [rollup/rollup#930]
I expected the following code to be unchanged by rollup:

```js
var test = 0, result = 0;
if (test === 0) ++result;
console.log(result);
```

However, the output was surprising:

```js
var result = 0;
console.log(result);
```

It appears to think it knows that `test === 0` is false, even though it’s obviously true here. And actually, `test !== 0` produces the same result. And it appears to likewise assume that non-deterministic expressions in if clauses are also false, as in this example, which generates the same output:

```js
var result = 0;
if (Math.random() > 0.5) ++result;
console.log(result);
```

**Comment by Rich-Harris** [comment]
Yikes! Am AFK so can't check, but I suspect it's mistaken about `result` not being used - only time it will remove an if statement based on the test alone is if there are no moving parts (ie just literals and comparisons between literals).

**Comment by eventualbuddha** [comment]
I looked into this a bit. It seems it's due to rollup asking the `UpdateExpression` (`++result`) whether it has effects. Rollup decides that it doesn't because it sees the declaration for `result` hasn't yet been activated, but that's only because it hasn't called `run` on `console.log(result)` yet. This is only an issue with top-level statements as this works fine:

```js
{
var result = 0;
if (!result) result++;
console.log(result);
}
```

It seems to me either that `hasEffects` needs to do more work to determine whether it really is used or there should be a complete `run` pass before asking about `hasEffects`.

**Comment by Rich-Harris** [comment]
`hasEffects` is working as expected, though the naming of it is perhaps slightly misleading. There are basically two phases to the tree-shaking – a top-down phase (which uses `hasEffects`) and a bottom-up phase, which re-checks assignment/update/call expressions to see if they've changed their minds about whether or not they have effects now that we know more about the bundle (i.e. once we see the `console.log(result)` call and know that `result` is necessary).

The second phase repeats until no new previously unmarked code is marked for inclusion (I guess this is the 'shaking' part). Previously, Rollup just kept doing the top-down thing, which would mean walking the AST of every module. The bottom-up approach is intended to be more efficient. It just turns out it was missing the 'up' part of 'bottom-up' – the nearest statement would be marked, but parent statements (such as an if statement that didn't independently have effects that were already known, as in this case) would not be.

Fixed in 0.35.14. Thanks!

**Comment by mbostock** [comment]
Nice work. Thanks for the fix! 👍
